test/RegisterSpec.hs

```diff
@@ -2,40 +2,94 @@
 allOnes = fromList [I, I ..]
 
 testAnd :: KnownNat n => Register n -> Expectation
 testAnd r = do
     Register.and r r `shouldBe` r
-    Register.and r (Register.not r) `shouldBe` fromList [O, O ..]
-    -- TODO(thies): and with allZeros and allOnes, implement other tests
+    Register.and r (Register.not r) `shouldBe` allZeros
+    Register.and r allZeros `shouldBe` allZeros
 
 testOr :: KnownNat n => Register n -> Expectation
 testOr r = do
     Register.or r r `shouldBe` r
-    Register.or r (Register.not r) `shouldBe` fromList [I, I ..]
+    Register.or r (Register.not r) `shouldBe` allOnes
+
+testOrNotZero :: KnownNat n => Register n -> Expectation
+testOrNotZero r
+    | r == allZeros = Register.or r allZeros `shouldBe` allZeros
+    | otherwise = Register.or r allZeros `shouldNotBe` allZeros
+
+testNand :: KnownNat n => Register n -> Register n -> Expectation
+testNand r1 r2 = Register.nand r1 r2 `shouldBe` (Register.not $ Register.and r1 r2)
+
+testNor :: KnownNat n => Register n -> Register n -> Expectation
+testNor r1 r2 = Register.nor r1 r2 `shouldBe` (Register.not $ Register.or r1 r2)
+
+testXor :: KnownNat n => Register n -> Expectation
+testXor r = do
+    Register.xor r (Register.not r) `shouldBe` allOnes
+    Register.xor r r `shouldBe` allZeros
+
+testXnor :: KnownNat n => Register n -> Register n -> Expectation
+testXnor r1 r2 = Register.xnor r1 r2 `shouldBe` (Register.not $ Register.xor r1 r2)
+
+impliesAllZeros :: KnownNat n => Register n -> Register n -> Expectation
+impliesAllZeros r1 r2
+    | r2 == allZeros = Register.implies allOnes r2 `shouldBe` allZeros
+    | otherwise = Register.implies r1 r2 `shouldNotBe` allZeros
+
+impliesAllOnes :: KnownNat n => Register n -> Expectation
+impliesAllOnes r = Register.implies r allOnes `shouldBe` allOnes
 
 spec :: Spec
 spec = do
-  describe "not" $ do
-    it "changes each Bit" $ property $ notChangesEachBit @4
-    it "is its own inverse" $ property $ notIsItsOwnInverse @4
-  describe "and" $ undefined
-  describe "or" $ undefined
+    describe "not" $ do
+        it "changes each Bit" $ property $ notChangesEachBit @4
+        it "is its own inverse" $ property $ notIsItsOwnInverse @4
+    describe "and" $
+        it "r and r = r; r and (not r) = zero, r and zero = zero" $
+            property $
+                testAnd @4
+    describe "or" $ do
+        it "r or r == r; r or not r == all ones" $ property $ testOr @4
+        it "is only zero if both inputs are zero" $ property $ testOrNotZero @4
+    describe "nand" $
+        it "is the opposite of and" $
+            property $
+                testNand @4
+    describe "nor" $
+        it "is the opposite of or" $
+            property $
+                testNor @4
+    describe "xor" $
+        it "r xor r = zero, r xor not r = all ones" $
+            property $
+                testXor @4
+    describe "xnor" $
+        it "is the opposite of xor" $
+            property $
+                testXnor @4
```

Comment:
These are terrible descriptions. They should say why the test exists, not what it does.

Reply:
How to write non-terrible descriptions for tests that test for more than one property? Should they be multiple lines? If yes, why not write one test per property?
And isn't the definition of nand, nor and xnor that they're "abbreviations" for not and, not or, not xor, so the functions are sufficiently tested if they behave that way?

Reply:
> How to write non-terrible descriptions for tests that test for more than one property

Maybe something along the lines of «works as expected»... I'm also not great at writing these. :sweat_smile: 

Maybe @fricklerhandwerk or @joto314 can help? 

> isn't the definition of nand, nor and xnor that they're "abbreviations" for not and, not or, not xor

The purpose of a test description is *not* to describe the function that is being tested. Such a description should go into a docstring above the function, but it is not the purpose of the test.

> so the functions are sufficiently tested if they behave that way

Sure, agreed.

Reply:
> These are terrible descriptions. They should say why the test exists, not what it does.

@kirschmet2 since I can't tell either why exactly these tests exist at all, I don't really have a suggestion. Both the functions as well as the tested properties are so trivial as to be evidently correct. They are literally the bitwise operations applied to the each bit in the vector. Maybe this is the only property we actually have to test, if at all.

Reply:
Maybe it makes sense to see these tests as testing register-level semantics in isolation from bitwise operations.

e.g. we expect that `and r r` will be `r`, which works for both bitwise and register `and`. At the same time it also exhaustively tests the register’s `Eq` instance.


Maybe the bitwise and register tests should be exactly the same, and so far I’m liking these register tests more than the bitwise ones (which are a bit similar to truth tables atm).